src/invidious/helpers/youtube_api.cr

```diff
@@ -2,120 +2,383 @@
+  extend self
 
-####################################################################
-# make_youtube_api_context(region)
-#
-# Return, as a Hash, the "context" data required to request the
-# youtube API endpoints.
-#
-def make_youtube_api_context(region : String | Nil) : Hash
-  return {
-    "client" => {
-      "hl"            => "en",
-      "gl"            => region || "US", # Can't be empty!
-      "clientName"    => "WEB",
-      "clientVersion" => HARDCODED_CLIENT_VERS,
+  # Enumerate used to select one of the clients supported by the API
+  enum ClientType
+    Web     = 0
+    Android = 1
+  end
+
+  # List of hard-coded values used by the different clients
+  HARDCODED_CLIENTS = {
+    ClientType::Web => {
+      name:    "WEB",
+      version: "2.20210721.00.00",
+      api_key: "AIzaSyAO_FJ2SlqU8Q4STEHLGCilw_Y9_11qcW8",
+    },
+    ClientType::Android => {
+      name:    "ANDROID",
+      version: "16.20.35",
+      api_key: "AIzaSyA8eiZmM1FaDVjRy-df2KTyQ_vz_yYM39w",
     },
   }
-end
 
-####################################################################
-# request_youtube_api_browse(continuation)
-# request_youtube_api_browse(browse_id, params, region)
-#
-# Requests the youtubei/v1/browse endpoint with the required headers
-# and POST data in order to get a JSON reply in english that can
-# be easily parsed.
-#
-# The region can be provided, default is US.
-#
-# The requested data can either be:
-#
-#  - A continuation token (ctoken). Depending on this token's
-#    contents, the returned data can be comments, playlist videos,
-#    search results, channel community tab, ...
-#
-#  - A playlist ID (parameters MUST be an empty string)
-#
-def request_youtube_api_browse(continuation : String)
-  # JSON Request data, required by the API
-  data = {
-    "context"      => make_youtube_api_context("US"),
-    "continuation" => continuation,
-  }
+  ####################################################################
+  # struct ClientConfig
+  #
+  # Data structure used to pass a client configuration to the different
+  # API endpoints handlers.
+  #
+  # Use case examples:
+  #
+  # ```
+  # # Get Norwegian search results
+  # conf_1 = ClientConfig.new(region: "NO")
+  # YoutubeAPI::search("Kollektivet", params: "", client_config: conf_1)
+  #
+  # # Use the Android client to request video streams URLs
+  # conf_2 = ClientConfig.new(client_type: ClientType::Android)
+  # YoutubeAPI::player(video_id: "dQw4w9WgXcQ", client_config: conf_2)
+  #
+  # # Proxy request through russian proxies
+  # conf_3 = ClientConfig.new(proxy_region: "RU")
+  # YoutubeAPI::next({video_id: "dQw4w9WgXcQ"}, client_config: conf_3)
+  # ```
+  #
+  struct ClientConfig
+    # Type of client to emulate (Web or Android).
+    # See `enum ClientType` and `HARDCODED_CLIENTS`.
+    property client_type : ClientType
 
-  return _youtube_api_post_json("/youtubei/v1/browse", data)
-end
+    # Region to provide to youtube, e.g to alter search results
```

Comment:
I think this should specify that the region parameter should be an ISO county code.

Reply:
Well, I'm not sure about that. This was my first guess, but I'm not sure how to verify that.